packages/medusa/src/repositories/price-list.ts

```diff
@@ -1,5 +1,96 @@
-import { EntityRepository, Repository } from "typeorm"
+import { groupBy, map } from "lodash"
+import {
+  Brackets,
+  EntityRepository,
+  FindManyOptions, Repository
+} from "typeorm"
 import { PriceList } from "../models/price-list"
+import { CustomFindOptions } from "../types/common"
+
+type PriceListFindOptions = CustomFindOptions<PriceList, 'status' | 'type'>
 
 @EntityRepository(PriceList)
-export class PriceListRepository extends Repository<PriceList> {}
+export class PriceListRepository extends Repository<PriceList> {
+  public async getFreeTextSearchResultsAndCount(
+    q: string,
+    options: PriceListFindOptions = { where: {} },
+    relations: (keyof PriceList)[] = []
+  ): Promise<[PriceList[], number]> {
+    options.where = options.where ?? {}
+    let qb = this.createQueryBuilder("price_list")
+      .leftJoinAndSelect("price_list.customer_groups", "customer_group")
+      .select(["price_list.id"])
+      .where(options.where)
+      .andWhere(
+        new Brackets((qb) => {
+          qb.where(`price_list.description ILIKE :q`, { q: `%${q}%` })
+            .orWhere(`price_list.name ILIKE :q`, { q: `%${q}%` })
+            .orWhere(`customer_group.name ILIKE :q`, { q: `%${q}%` })
+        })
+      )
+      .skip(options.skip)
+      .take(options.take)
+
+    const [results, count] = await qb.getManyAndCount()
+
+    const price_lists = await this.findWithRelations(
+      relations,
+      results.map((r) => r.id)
+    )
+
+    return [price_lists, count]
+  }
+
+  public async findWithRelations(
+    relations: (keyof PriceList)[] = [],
+    idsOrOptionsWithoutRelations:
+      | Omit<FindManyOptions<PriceList>, "relations">
+      | string[] = {}
+  ): Promise<PriceList[]> {
+    let entities
+    if (Array.isArray(idsOrOptionsWithoutRelations)) {
+      entities = await this.findByIds(idsOrOptionsWithoutRelations)
+    } else {
+      entities = await this.find(idsOrOptionsWithoutRelations)
+    }
+    
+    const groupedRelations: Record<string, string[]> = {}
+    for (const relation of relations) {
+      const [topLevel] = relation.split(".")
+      if (groupedRelations[topLevel]) {
+        groupedRelations[topLevel].push(relation)
+      } else {
+        groupedRelations[topLevel] = [relation]
+      }
+    }
+    
+    const entitiesIds = entities.map(({ id }) => id)
+    const entitiesIdsWithRelations = await Promise.all(
+      Object.values(groupedRelations).map((relations: string[]) => {
+        return this.findByIds(entitiesIds, {
+          select: ["id"],
+          relations: relations as string[],
+        })
+      })
+    ).then(entitiesIdsWithRelations => entitiesIdsWithRelations.flat())
```

Comment:
Careful, the flat method only flatten the number of level specified. If none then it flatten one level, is that the intended behaviour ? Just to be sure 😅

Reply:
I think so. Before the requested change, we passed lodash's [flatten](https://lodash.com/docs/4.17.15#flatten) which also flattens a single level, hence why I used `flat` with the default argument. Also from my understanding, I don't think we'll exceed 1 level of depth no?

Reply:
We shouldn't indeed, but just wanted to be sure I al not missing something 🤣